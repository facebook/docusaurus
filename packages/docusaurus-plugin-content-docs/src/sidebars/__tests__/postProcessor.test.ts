/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  postProcessSidebars,
  type SidebarPostProcessorParams,
} from '../postProcessor';

describe('postProcess', () => {
  it('transforms category without subitems', () => {
    const processedSidebar = postProcessSidebars(
      {
        sidebar: [
          {
            type: 'category',
            label: 'Category',
            link: {
              type: 'generated-index',
              slug: 'generated/permalink',
            },
            items: [],
          },
          {
            type: 'category',
            label: 'Category 2',
            link: {
              type: 'doc',
              id: 'doc ID',
            },
            items: [],
          },
        ],
      },
      {
        sidebarOptions: {sidebarCollapsed: true, sidebarCollapsible: true},
        version: {path: 'version'},
        drafts: [],
      } as unknown as SidebarPostProcessorParams,
    );

    expect(processedSidebar).toMatchSnapshot();

    expect(() => {
      postProcessSidebars(
        {
          sidebar: [
            {
              type: 'category',
              label: 'Bad category',
              items: [],
            },
          ],
        },
        {
          sidebarOptions: {sidebarCollapsed: true, sidebarCollapsible: true},
          version: {path: 'version'},
          drafts: [],
        } as unknown as SidebarPostProcessorParams,
      );
    }).toThrowErrorMatchingInlineSnapshot(
      `"Sidebar category Bad category has neither any subitem nor a link. This makes this item not able to link to anything."`,
    );
  });

  it('corrects collapsed state inconsistencies', () => {
    expect(
      postProcessSidebars(
        {
          sidebar: [
            {
              type: 'category',
              label: 'Category',
              collapsed: true,
              collapsible: false,
              items: [{type: 'doc', id: 'foo'}],
            },
          ],
        },

        {
          sidebarOptions: {sidebarCollapsed: true, sidebarCollapsible: true},
          version: {path: 'version'},
          drafts: [],
        } as unknown as SidebarPostProcessorParams,
      ),
    ).toMatchSnapshot();

    expect(
      postProcessSidebars(
        {
          sidebar: [
            {
              type: 'category',
              label: 'Category',
              collapsed: true,
              items: [{type: 'doc', id: 'foo'}],
            },
          ],
        },

        {
          sidebarOptions: {sidebarCollapsed: false, sidebarCollapsible: false},
          version: {path: 'version'},
          drafts: [],
        } as unknown as SidebarPostProcessorParams,
      ),
    ).toMatchSnapshot();

    expect(
      postProcessSidebars(
        {
          sidebar: [
            {
              type: 'category',
              label: 'Category',
              items: [{type: 'doc', id: 'foo'}],
            },
          ],
        },

        {
          sidebarOptions: {sidebarCollapsed: true, sidebarCollapsible: false},
          version: {path: 'version'},
          drafts: [],
        } as unknown as SidebarPostProcessorParams,
      ),
    ).toMatchSnapshot();
  });

  it('filters draft items', () => {
    expect(
      postProcessSidebars(
        {
          sidebar: [
            {
              type: 'category',
              label: 'Category',
              items: [{type: 'doc', id: 'foo'}],
            },
            {
              type: 'category',
              label: 'Category',
              link: {
                type: 'doc',
                id: 'another',
              },
              items: [{type: 'doc', id: 'foo'}],
            },
          ],
        },
        {
          sidebarOptions: {sidebarCollapsed: true, sidebarCollapsible: true},
          version: {path: 'version'},
          drafts: [{id: 'foo'}],
        } as unknown as SidebarPostProcessorParams,
      ),
    ).toMatchSnapshot();
  });
});
