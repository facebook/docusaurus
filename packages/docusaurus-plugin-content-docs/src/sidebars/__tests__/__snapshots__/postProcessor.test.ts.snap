// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postProcess corrects collapsed state inconsistencies 1`] = `
{
  "sidebar": [
    {
      "collapsed": false,
      "collapsible": false,
      "items": [
        {
          "id": "foo",
          "type": "doc",
        },
      ],
      "label": "Category",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`postProcess corrects collapsed state inconsistencies 2`] = `
{
  "sidebar": [
    {
      "collapsed": false,
      "collapsible": false,
      "items": [
        {
          "id": "foo",
          "type": "doc",
        },
      ],
      "label": "Category",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`postProcess corrects collapsed state inconsistencies 3`] = `
{
  "sidebar": [
    {
      "collapsed": false,
      "collapsible": false,
      "items": [
        {
          "id": "foo",
          "type": "doc",
        },
      ],
      "label": "Category",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`postProcess filters draft items 1`] = `
{
  "sidebar": [
    {
      "id": "another",
      "label": "Category",
      "type": "doc",
    },
  ],
}
`;

exports[`postProcess transforms category without subitems 1`] = `
{
  "sidebar": [
    {
      "id": "doc ID",
      "label": "Category 2",
      "type": "doc",
    },
  ],
}
`;
